### ISSUE TITLE
Streaming not working when routing between Runnables in LCEL

### ISSUE THREAD

**Issue by daolmedo** [langchain-ai/langchain#11998]
### System Info

langchain_version: "0.0.306"
library: "langchain"
library_version: "0.0.306"
platform: "Windows-10-10.0.22621-SP0"
py_implementation: "CPython"
runtime: "python"
runtime_version: "3.9.0rc2"
sdk_version: "0.0.41"

### Who can help?

@hwchase17 

### Information

- [ ] The official example notebooks/scripts
- [X] My own modified scripts

### Related Components

- [X] LLMs/Chat Models
- [ ] Embedding Models
- [ ] Prompts / Prompt Templates / Prompt Selectors
- [ ] Output Parsers
- [ ] Document Loaders
- [ ] Vector Stores / Retrievers
- [ ] Memory
- [ ] Agents / Agent Executors
- [ ] Tools / Toolkits
- [ ] Chains
- [ ] Callbacks/Tracing
- [ ] Async

### Reproduction

Streaming the response of a simple route created among 2 Runnables doesn't work. When I stream each Runnable independelty, it streams perfectly. 

When I create a route for those Runnables to choose which one of them should be the next step, it doesn't stream the result. 

Here's the example:
```
# ----------------- Runnable 1 -----------------

DEFAULT_DOCUMENT_PROMPT = PromptTemplate.from_template(template="{page_content}")
def _combine_documents(docs, document_prompt = DEFAULT_DOCUMENT_PROMPT, document_separator="\n\n"):
  doc_strings = [format_document(doc, document_prompt) for doc in docs]
  return document_separator.join(doc_strings)
  
def _format_chat_history(chat_history: List[Tuple]) -> str:
  buffer = ""
    for dialogue_turn in chat_history:
      human = "Human: " + dialogue_turn[0]
      ai = "Assistant: " + dialogue_turn[1]
      buffer += "\n" + "\n".join([human, ai])
  return buffer
  
_inputs = RunnableMap({
  "standalone_question": {
    "question": lambda x: x["question"],
    "chat_history": lambda x: _format_chat_history(x['chat_history'])
  } | PromptFactory.CONDENSE_QUESTION_PROMPT | ChatOpenAI(temperature=0, streaming=True) | StrOutputParser(),
})
_context = {
  "context": itemgetter("standalone_question") | vector_retriever | _combine_documents,
  "question": lambda x: x["standalone_question"]
}
Runnable1 = _inputs | _context | PromptFactory.PROMPT | ChatOpenAI(temperature=0, streaming=True)

# ----------------- Stream the response: -----------------

for s in Runnable1.stream({"question": "Hello!", "chat_history": []}):
  print(s, end="", flush=True)
```
The above stream works
```
# ----------------- Runnable 2 -----------------

Runnable2 = RunnableMap({
  "response": {
    "question": lambda x: x["question"],
    "chat_history": lambda x: _format_chat_history(x['chat_history']),
    "context": lambda x: QueriesContext.get_result()
  } | PromptTemplate.from_template(PromptFactory.followup_context_template) | ChatOpenAI(temperature=0, streaming=True) | StrOutputParser(),
})

# ----------------- Stream the response: -----------------

for s in Runnable2.stream({"question": "Hello!", "chat_history": [], "context": ['initialvalue']}):
  print(s, end="", flush=True)
```
The above stream also works
```
def get_result(text):
  return ["newvalue"]
router_chain = PromptTemplate.from_template(PromptFactory.router_template_test) | ChatOpenAI(temperature=0, streaming=True) | StrOutputParser()

def route(info):
  if "runnable1" in info["topic"].lower():
    return Runnable1
  elif "runnable2" in info["topic"].lower():
    return Runnable2
  else:
    raise Exception("Invalid topic")

full_runnable_router_chain = {
  "topic": router_chain,
  "question": lambda x: x["question"],
  "chat_history": lambda x: x["chat_history"],
  "context": lambda x: x["context"]
} | RunnableLambda(route)

# ----------------- Stream the response: -----------------

for s in full_runnable_router_chain.stream({"question": "Hello!", "chat_history": [], "context": ['initialvalue']}):
  print(s.content, end="", flush=True)
```

The above stream doesn't stream

### Expected behavior

Stream from "full_runnable_router_chain"

**Comment by daolmedo** [comment]
I have tried the suggested modification and ran the same implementation without any changes to the original code posted. The Router Runnable outputs the response but it doesn't stream it.

**Comment by qtangs** [comment]
`RunnableLambda` on functions returning `Runnable` doesn't support streaming yet. I've tested a subclass with streaming, it works for me, let me know how it goes for your case.

```
from typing import Optional, Any, Iterator, AsyncIterator

from langchain.callbacks.manager \
    import (CallbackManagerForChainRun, AsyncCallbackManagerForChainRun)
from langchain.schema.runnable import RunnableLambda, RunnableConfig, Runnable
from langchain.schema.runnable.config \
    import (patch_config,
            call_func_with_variable_args, acall_func_with_variable_args)
from langchain.schema.runnable.utils import Input, Output


class StreamingRunnableLambda(RunnableLambda):
    def _stream(
            self,
            input: Input,
            run_manager: CallbackManagerForChainRun,
            config: RunnableConfig,
    ) -> Iterator[Output]:
        output = call_func_with_variable_args(self.func, input, config, run_manager)

        # If the output is a runnable, invoke it
        if isinstance(output, Runnable):
            recursion_limit = config["recursion_limit"]
            if recursion_limit <= 0:
                raise RecursionError(
                    f"Recursion limit reached when invoking {self} with input {input}."
                )
            output = output.stream(
                input,
                patch_config(
                    config,
                    callbacks=run_manager.get_child(),
                    recursion_limit=recursion_limit - 1,
                ),
            )
        return output

    def stream(
            self,
            input: Input,
            config: Optional[RunnableConfig] = None,
            **kwargs: Optional[Any],
    ) -> Iterator[Output]:
        if hasattr(self, "func"):
            return self._call_with_config(
                self._stream,
                input,
                self._config(config, self.func),
            )
        else:
            raise TypeError(
                "Cannot invoke a coroutine function synchronously."
                "Use `astream` instead."
            )
```

**Comment by daolmedo** [comment]
Thank you @qtangs, I have not been able to make your example run correctly. Can you please share a working example?

**Comment by qtangs** [comment]
Here you are :). This is an example adapted from your code, the most important change is to replace `RunnableLambda` with `StreamingRunnableLambda`:

```
if __name__ == '__main__':

    context = """
Following is a conversation between Clara, a climate advocate with a passion for environmental issues, 
and Pam, a peace activist dedicated to fostering harmony among humans.

Clara: You know, Pam, it's been on my mind lately – the interconnectedness between human relationships 
and our relationship with the environment.

Pam: Absolutely. I believe a peaceful world begins with understanding and respecting our interconnectedness, 
not just with each other but with the planet we call home.

Clara: That's precisely why I find our discussions so enriching. Climate change isn't just 
an environmental crisis; it's a reflection of how we relate to the world and each other.

Pam: It's fascinating how our treatment of the environment often mirrors our relationships. 
When there's neglect, exploitation, or lack of care, problems arise. It's the same with human connections.
    """

    router_chain = ChatPromptTemplate.from_template(
        template=
        "Determine if this question is related for `Clara` or `Pam`: `{question}`."
        "Return only the final answer."
    ) | ChatOpenAI(temperature=0, streaming=True) | StrOutputParser()

    # ----------------- Runnable 1 -----------------
    Runnable1 = RunnableMap({
        "response":
            {
                "question": lambda x: x["question"],
                "context": lambda x: context,
            } | ChatPromptTemplate.from_template(
                template="Given the following context: \n{context}\n\n."
                         "Answer this as Clara, start with `According to Clara, `: {question}"
            ) | ChatOpenAI(temperature=0, streaming=True) | StrOutputParser(),
    })

    # ----------------- Runnable 2 -----------------
    Runnable2 = RunnableMap({
        "response":
            {
                "question": lambda x: x["question"],
                "context": lambda x: context,
            } | ChatPromptTemplate.from_template(
                template="Given the following context: \n{context}\n\n."
                         "Answer this as Pam, start with `According to Pam, `: {question}"
            ) | ChatOpenAI(temperature=0, streaming=True) | StrOutputParser(),
    })

    # ----------------- Router Chain -----------------
    def route(info):
        if "clara" in info["topic"].lower():
            return Runnable1
        elif "pam" in info["topic"].lower():
            return Runnable2
        else:
            raise Exception("Invalid topic: " + info["topic"])

    full_runnable_router_chain = \
        {
            "topic": router_chain,
            "question": lambda x: x["question"],
            "chat_history": lambda x: x["chat_history"],
            "context": lambda x: x["context"],
        } | StreamingRunnableLambda(route)

    # ----------------- Stream the response: -----------------
    for s in full_runnable_router_chain.stream({
        "question": "What did Clara say about human relationships and our relationship with the environment?",
        "chat_history": [], "context": ['initialvalue']
    }):
        print(s["response"], end="", flush=True)
    for s in full_runnable_router_chain.stream({
        "question": "What did Pam say about human relationships and our relationship with the environment?",
        "chat_history": [], "context": ['initialvalue']
    }):
        print(s["response"], end="", flush=True)
```

**Comment by Talhavaival1** [comment]
> Here you are :). This is an example adapted from your code, the most important change is to replace `RunnableLambda` with `StreamingRunnableLambda`:
> 
> ```
> if __name__ == '__main__':
> 
>     context = """
> Following is a conversation between Clara, a climate advocate with a passion for environmental issues, 
> and Pam, a peace activist dedicated to fostering harmony among humans.
> 
> Clara: You know, Pam, it's been on my mind lately – the interconnectedness between human relationships 
> and our relationship with the environment.
> 
> Pam: Absolutely. I believe a peaceful world begins with understanding and respecting our interconnectedness, 
> not just with each other but with the planet we call home.
> 
> Clara: That's precisely why I find our discussions so enriching. Climate change isn't just 
> an environmental crisis; it's a reflection of how we relate to the world and each other.
> 
> Pam: It's fascinating how our treatment of the environment often mirrors our relationships. 
> When there's neglect, exploitation, or lack of care, problems arise. It's the same with human connections.
>     """
> 
>     router_chain = ChatPromptTemplate.from_template(
>         template=
>         "Determine if this question is for `Clara` or `Pam`: `{question}`."
>         "Return only the final answer."
>     ) | ChatOpenAI(temperature=0, streaming=True) | StrOutputParser()
> 
>     # ----------------- Runnable 1 -----------------
>     Runnable1 = RunnableMap({
>         "response":
>             {
>                 "question": lambda x: x["question"],
>                 "context": lambda x: context,
>             } | ChatPromptTemplate.from_template(
>                 template="Given the following context: \n{context}\n\n."
>                          "Answer this as Clara, start with `According to Clara, `: {question}"
>             ) | ChatOpenAI(temperature=0, streaming=True) | StrOutputParser(),
>     })
> 
>     # ----------------- Runnable 2 -----------------
>     Runnable2 = RunnableMap({
>         "response":
>             {
>                 "question": lambda x: x["question"],
>                 "context": lambda x: context,
>             } | ChatPromptTemplate.from_template(
>                 template="Given the following context: \n{context}\n\n."
>                          "Answer this as Pam, start with `According to Pam, `: {question}"
>             ) | ChatOpenAI(temperature=0, streaming=True) | StrOutputParser(),
>     })
> 
>     # ----------------- Router Chain -----------------
>     def route(info):
>         if "clara" in info["topic"].lower():
>             return Runnable1
>         elif "pam" in info["topic"].lower():
>             return Runnable2
>         else:
>             raise Exception("Invalid topic: " + info["topic"])
> 
>     full_runnable_router_chain = \
>         {
>             "topic": router_chain,
>             "question": lambda x: x["question"],
>             "chat_history": lambda x: x["chat_history"],
>             "context": lambda x: x["context"],
>         } | StreamingRunnableLambda(route)
> 
>     # ----------------- Stream the response: -----------------
>     for s in full_runnable_router_chain.stream({
>         "question": "What did Clara say about human relationships and our relationship with the environment?",
>         "chat_history": [], "context": ['initialvalue']
>     }):
>         print(s["response"], end="", flush=True)
>     for s in full_runnable_router_chain.stream({
>         "question": "What did Pam say about human relationships and our relationship with the environment?",
>         "chat_history": [], "context": ['initialvalue']
>     }):
>         print(s["response"], end="", flush=True)
> ```

Thank you so much , this and the previous answer combined give a proper streaming response for runnableBranch thank you.

**Comment by Talhavaival1** [comment]
@qtangs  , could please provide a implementation of the astream method as well since currently astream is not working when I try to stream it via fastapi , but the stream method is working correctly.

**Comment by qtangs** [comment]
@Talhavaival1, you can find the astream implementation in this PR: https://github.com/langchain-ai/langchain/pull/14794

**Comment by nfcampos** [comment]
Fixed in #14794
